docs/posts/ibisml/index.qmd

```diff
@@ -0,0 +1,897 @@
+X_train_transformed = last_mile_preprocessing.transform(X_train)
+X_test_transformed = last_mile_preprocessing.transform(X_test)
+print(f"train data shape = {X_train_transformed.shape}")
+print(f"test data shape = {X_test_transformed.shape}")
+```
+
+Let's define a neural network classifier using pytorch and pytorch lighting:
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to define a torch classifier"
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data import DataLoader, TensorDataset
+import pytorch_lightning as pl
+from pytorch_lightning import Trainer
+
+
+class NeuralNetClassifier(pl.LightningModule):
+    def __init__(self, input_dim, hidden_dim=8, output_dim=1):
+        super().__init__()
+        self.model = nn.Sequential(
+            nn.Linear(input_dim, hidden_dim),
+            nn.ReLU(),
+            nn.Linear(hidden_dim, output_dim),
+        )
+        self.loss = nn.BCEWithLogitsLoss()
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x):
+        return self.model(x)
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
+        y_hat = self(x)
+        loss = self.loss(y_hat.view(-1), y)
+        self.log("train_loss", loss)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        x, y = batch
+        y_hat = self(x)
+        loss = self.loss(y_hat.view(-1), y)
+        self.log("val_loss", loss)
+        return loss
+
+    def configure_optimizers(self):
+        return optim.Adam(self.parameters(), lr=0.001)
+
+    def predict_proba(self, x):
+        self.eval()
+        with torch.no_grad():
+            x = x.to(self.device)
+            return self.sigmoid(self(x))
+
+
+# Initialize your Lightning Module
+nn_classifier = NeuralNetClassifier(input_dim=X_train_transformed.shape[1])
+
+y_train_array = y_train.to_pandas().to_numpy().astype(np.float32)
+x_train_tensor = torch.from_numpy(X_train_transformed)
+y_train_tensor = torch.from_numpy(y_train_array)
+train_dataset = TensorDataset(x_train_tensor, y_train_tensor)
+
+y_test_array = y_test.to_pandas().to_numpy().astype(np.float32)
+X_test_tensor = torch.from_numpy(X_test_transformed)
+y_test_tensor = torch.from_numpy(y_test_array)
+val_dataset = TensorDataset(X_test_tensor, y_test_tensor)
+
+train_loader = DataLoader(train_dataset, batch_size=32, shuffle=False)
+val_loader = DataLoader(val_dataset, batch_size=32, shuffle=False)
+
+# Initialize a Trainer
+trainer = Trainer(max_epochs=2)
+print(nn_classifier)
+```
+
+Let's train the classifier:
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to train the pytorch classifier"
+# Train the model
+trainer.fit(nn_classifier, train_loader, val_loader)
+```
+
+Let's use the trained model to make a prediction:
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to predict using the trained pytorch classifier"
+y_pred = nn_classifier.predict_proba(X_test_tensor[:10])
+y_pred
+```
+
```

Comment:
Would love to see some sort of conclusion/takeaways? One thing I recall you mentioning is that the code seemed cleaner/simpler. It might be interesting to point out that we're scaling the preprocessing on DuckDB (a selling point).